### ISSUE TITLE
NPE in DirectoryServerResource if the URL contains a query

### ISSUE THREAD

**Issue by ralphvanetten** [restlet/restlet-framework-java#1206]
When serving files using the `Directory` resource in Restlet 2.3.7 I get an NPE if the URL contains a query. Before the `DirectoryServerResource` refactoring this worked fine and `DirectoryServerResource` would just ignore the query.

The NPE is caused by https://github.com/restlet/restlet-framework-java/blob/2.3.7/modules/org.restlet/src/org/restlet/engine/local/DirectoryServerResource.java#L199 assigning `null` to `this.relativePart` which is then accessed on line 203.

I think it is caused by `Reference#getRemainingRef()` which seems to return `null` when there is a query argument: 
The url `http://myapplication/fichier.txt` gives `fichier.txt` as remaining ref 
The url `http://myapplication/fichier.txt?x=y` gives `null` as remaining ref while I would expect this to also be `fichier.txt` or `fichier.txt?x=y`. (this was observed using the test case described below)

Stacktrace:
```
java.lang.NullPointerException
        at org.restlet.engine.local.DirectoryServerResource.doInit(DirectoryServerResource.java:205)
        at org.restlet.resource.Resource.init(Resource.java:767)
        at org.restlet.resource.Finder.handle(Finder.java:233)
        at org.restlet.resource.Directory.handle(Directory.java:239)
```

Patch for the test case which reproduces the NPE:
```
diff --git a/modules/org.restlet.test/src/org/restlet/test/resource/DirectoryTestCase.java b/modules/org.restlet.test/src/org/restlet/test/resource/DirectoryTestCase.java
index f77c8a8..9a39fac 100644
--- a/modules/org.restlet.test/src/org/restlet/test/resource/DirectoryTestCase.java
+++ b/modules/org.restlet.test/src/org/restlet/test/resource/DirectoryTestCase.java
@@ -481,6 +481,14 @@ public class DirectoryTestCase extends RestletTestCase {
         assertEquals(SUCCESS_OK, response.getStatus());
         assertTrue(response.getEntityAsText().equals("this is test 3b"));
 
+
+        response = new TestRequest(this.baseFileUrl + "?x=y")
+                .baseRef(this.webSiteURL)
+                .handle(GET);
+        assertEquals(SUCCESS_OK, response.getStatus());
+        assertTrue(response.getEntityAsText().equals("this is test 3b"));
+
+
         // Test 5 : add a new representation of the same base file
         response = new TestRequest(this.baseFileUrlEn)
                 .baseRef(this.webSiteURL)
diff --git a/modules/org.restlet.test/src/org/restlet/test/resource/ResourceTestSuite.java b/modules/org.restlet.test/src/org/restlet/test/resource/ResourceTestSuite.java
index b36e98c..99f1360 100644
--- a/modules/org.restlet.test/src/org/restlet/test/resource/ResourceTestSuite.java
+++ b/modules/org.restlet.test/src/org/restlet/test/resource/ResourceTestSuite.java
@@ -32,7 +32,7 @@ public class ResourceTestSuite {
     public static Test suite() {
         TestSuite suite = new TestSuite("Resource package");
         // $JUnit-BEGIN$
-        // suite.addTestSuite(DirectoryTestCase.class);
+        suite.addTestSuite(DirectoryTestCase.class);
 
         suite.addTestSuite(AnnotatedResource01TestCase.class);
         suite.addTestSuite(AnnotatedResource02TestCase.class);
```

**Comment by thboileau** [comment]
Hello @ralphvanetten 

thanks a lot for reporting this issue. A fix is available in the the 2.3 branch.
Do you need a new version? We can release one, if needed at the beginning of next week.

**Comment by ralphvanetten** [comment]
Many thanks for fixing it! Currently I have a work around in place so a new release is not urgent at all.

**Comment by PSpSynedra** [comment]
We ran into the same issue using Restlet-2.3.7. As a workaround we simply removed the trailing "?{rq}" from an applied redirect pattern as we currently don't need/use it.

@thboileau: Will the fix make it into Restlet-2.3.8? I am not sure if I read the commit details of c900125 correctly.

**Comment by thboileau** [comment]
Hello @PSpSynedra, the fix is part of the 2.3.8 and it works well (I've just tested it)
